### ISSUE TITLE
windows-latest workflows will use Windows Server 2022

### ISSUE THREAD

**Issue by miketimofeev** [actions/virtual-environments#4856]
### Breaking changes

Windows Server 2022 is ready to be the default version for the "windows-latest" label in GitHub Actions and Azure DevOps.

### Target date

This change will be rolled out over a period of several weeks beginning on January, 17. We plan to complete the migration by March, 6.

### The motivation for the changes

Starting from November 2021 Windows Server 2022 with Visual Studio 2022 image is generally available for all customers. We have monitored customer feedback to improve the Windows Server 2022 image stability and now we are ready to set it as the latest.

### Possible impact
~~⚠️  [CodeQL](https://github.com/github/codeql-action) does not currently support analysis of compiled languages on Windows Server 2022 ⚠️~~ Support for Windows Server 2022 has been added.

The Windows Server 2022 image has a different set of software than Windows Server 2019. The most significant changes are listed in the table below:
|Tool name|Windows-2019|Windows-2022|Notes|
|-|:-:|-|-|
|Go lang|1.15.*|1.16.*| Go lang version bumped to the more recent one |
|Python|3.7.*|3.9.*| Python version bumped to the more recent one |
|Ruby|2.5.*| 3.0.* | Ruby version bumped to the more recent one |
|Google Cloud SDK|latest| - | The software was removed due to maintenance concerns, consider using the official action https://github.com/google-github-actions/setup-gcloud |
|Innosetup|latest| - | The software was removed due to its low popularity and maintenance concerns |
|Parcel|latest| - | The software was removed due to its low popularity and maintenance concerns |
|Cloud Foundry CLI |latest| - | The software was removed due to its low popularity and maintenance concerns |
|Java|8(default)<br>11<br>13<br>17|8(default)<br>11<br>17| Due to our software guidelines we left only LTS versions on the image. If you need to use other versions of Java, consider using [setup-java](https://github.com/actions/setup-java) and [Java tool installer](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/java-tool-installer?view=azure-devops) tasks |
|Python toolcache|2.7<br>3.5<br>3.6<br>3.7<br>3.8<br>3.9<br>3.10|3.7<br>3.8<br>3.9<br>3.10| EOL python versions were removed |
|PyPy toolcache|2.7<br>3.6<br>3.7<br>3.8|2.7<br>3.7<br>3.8| Version 3.6 was deprecated as it won't receive any updates |
|Ruby toolcache|2.4<br>2.5(default)<br>2.6<br>2.7<br>3.0|2.7<br>3.0(default)| Left only two recent versions |
|MySQL| 5.7.* | 8.0.* | The version was updated to the latest major one |
|.NET | 2.1<br>3.1<br>5.0| 3.1<br>5.0 | Version 2.1 is EOL and was removed |
|Windows Driver Kit | 10.0.* | - | VS2022 is not yet supported by the WDK |
|Az powershell module| 1.0.0.zip<br>1.6.0.zip<br>2.3.2.zip<br>2.6.0.zip<br>3.1.0.zip<br>3.5.0.zip<br>3.8.0.zip<br>4.3.0.zip<br>4.4.0.zip<br>4.7.0.zip<br>5.5.0.zip<br>5.9.0.zip<br>6.5.0 | 6.5 | Left only the most recent 6.* release |
|Android SDK build-tools and platforms | 19 - 32 | 27 - 32 | Android Platforms and build tools < 27.0 were removed from image. Consider installation in runtime using sdkmanager |

### Virtual environments affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [X] Windows Server 2019
- [X] Windows Server 2022

### Mitigation ways

If you see any issues with your workflows during this transition period:

- File an issue in this repository
- Switch back to windows-2019 by specifying the `windows-2019` label in your workflow. We will continue to support Windows Server 2019 image.

**Comment by tt2468** [comment]
Are there any known actions to install innosetup?

**Comment by BLUE-DEVIL1134** [comment]
> Are there any known actions that we can use to install innosetup?

Nah, I don't think that's possible from actions.

**Comment by miketimofeev** [comment]
@tt2468 innosetup can be installed relatively fast in runtime. Please feel free to ask for help if needed

**Comment by MellyDawgKY** [comment]
Why this was not done on the day of formally releasing visual studio 2022?

I always assumed Microsoft would follow the literal definition of 'latest' -  silly me.

**Comment by neman** [comment]
<img width="493" alt="image" src="https://user-images.githubusercontent.com/350314/150398675-f58a3fc9-643e-4dfc-bb24-5fbeb6424321.png">

What about LTS .NET 6.0?

**Comment by miketimofeev** [comment]
@neman 6.0 is bundled in VS2022, so it's available on the image but not installed via part of .NET installation script

**Comment by theolivenbaum** [comment]
@miketimofeev  Shouldn't the list be updated to reflect net6.0 being available?

**Comment by remyjette** [comment]
I'm seeing this warning
> ##[warning]windows-latest pipelines will use windows-2022 soon. For more details, see https://github.com/actions/virtual-environments/issues/4856

In all of my Windows pipelines, but none of my pipelines use `windows-latest` - they all use `windows-2019`.

**Comment by mikeKuester** [comment]
> Why this was not done on the day of formally releasing visual studio 2022?

Same for me! mairaw mentioned the issue I opend. I think that the name "windows-latest" is totally misleading. Something like "..-stable" would fit much better to the reality.

**Comment by miketimofeev** [comment]
> I'm seeing this warning
> 
> > ##[warning]windows-latest pipelines will use windows-2022 soon. For more details, see #4856
> 
> In all of my Windows pipelines, but none of my pipelines use `windows-latest` - they all use `windows-2019`.

Sorry for the inconvenience, this is a known issue and will be fixed next week.

**Comment by catthehacker** [comment]
> Something like "..-stable" would fit much better to the reality.

`-stable` implies that it is more "stable" than other `windows-*` labels, which is not and would be even more misleading

**Comment by mikeKuester** [comment]
Ok, good point, do you have a better idea?

With "-latest" I expect the "latest bits and bytes". Not that several month after the release of the actual .NET 6 it's not included. :(

**Comment by catthehacker** [comment]
the label `windows-*` pertains to Windows version, not software it includes
since `windows-2019` is Windows Server 2019 and `windows-2022` is Windows Server 2022, naturally, `windows-latest` is latest GA version of Windows Server

**Comment by stap123** [comment]
FWIW

I agree `windows-latest` implies the most recent GA versions of stuff. This would mean that as soon as `windows-2022` was GA then the `-latest` version should have switched over the same day (or close to it) not 6 months later (windows 2022 was GA in Sep).

When you take into consideration that .NET 6 was not available on the `-2019` image on release day either it was a bit frustrating being encouraged to update to the latest .NET asap and then having to find out that it actually required changes to pipelines because `-latest` doesn't mean latest 🥲 Acknowledging that the solution was to pin to `-2022` so not the end of the world it still meant confusion and frustration when the expectation was for the latest bits to be available.

There was a similar thing with Azure Functions last year with a many month lag between .NET 5 and support for it in functions. They worked to ensure that .NET 6 was supported on day 1 when that came out in Nov. Maybe there's a lesson to be learned there? 

@mikeKuester 
Another possible option would be to simply not offer a `windows-latest` and only offer pinned versions then people would have to take control of their own versions. I accept that makes it slightly more painful for people managing windows versions and many pipelines but it does remove the ambiguity.

**Comment by catthehacker** [comment]
> I agree `windows-latest` implies the most recent GA versions of stuff. This would mean that as soon as `windows-2022` was GA then the `-latest` version should have switched over the same day (or close to it) not 6 months later (windows 2022 was GA in Sep).

It takes time to work out all the quirks of new software when it's released to not break environment. Switching over to `windows-2022` on day 0 would break a lot of pipelines/workflows as well as it would require for GHA team to have it available on that time in Azure environment. I'll admit it is my fault that I didn't put `*` with clarification that latest OS requires additional work, thorough testing and customer feedback before it can be released "in the wild" and all that combined takes time.

> When you take into consideration that .NET 6 was not available on the `-2019` image on release day either it was a bit frustrating being encouraged to update to the latest .NET asap and then having to find out that it actually required changes to pipelines because `-latest` doesn't mean latest

The correct way is to use `dotnet` installation task ([ADO](https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/DotNetCoreInstallerV1) | [GHA](https://github.com/actions/setup-dotnet)) and not rely on specific bits installed on the host as much as possible.

**Comment by stap123** [comment]
> It takes time to work out all the quirks of new software when it's released to not break environment. Switching over to `windows-2022` on day 0 would break a lot of pipelines/workflows as well as it would require for GHA team to have it available on that time in Azure environment. I'll admit it is my fault that I didn't put `*` with clarification that latest OS requires additional work, thorough testing and customer feedback before it can be released "in the wild" and all that combined takes time. What I originally meant by `GA` is GA in sense of OS vendor **and** GitHub Actions.

Totally understand it takes time to ensure everything works it was more an observation of the fact that `-latest` implies a little risk because you're asking for the latest 😃 Thanks for the clarification though.

> The correct way is to use `dotnet` installation task ([ADO](https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/DotNetCoreInstallerV1?rgh-link-date=2022-01-21T11%3A33%3A23Z) | [GHA](https://github.com/actions/setup-dotnet)) and not rely on specific bits installed on the host as much as possible.

That's a fair point, on the flip side would it not make sense to have no .NET versions available on the image by default so that it's not possible for people to do it wrong? This would force people down the happy path when creating pipelines.

Keep up the good work 🍾

**Comment by MarcoWesterink** [comment]
> @neman 6.0 is bundled in VS2022, so it's available on the image but not installed via part of .NET installation script

Still might be good to add to the table. It confused me too.

**Comment by catthehacker** [comment]
> That's a fair point, on the flip side would it not make sense to have no .NET versions available on the image by default so that it's not possible for people to do it wrong? This would force people down the happy path when creating pipelines.

Some versions of software that can be installed through `actions/setup-*` is kept in the image because of popularity (of version and not only) and (not sure if all since I didn't check all actions but) `setup-*` tasks will match against requested version and if it matches, it will skip installation (e.g. task requires version '5.0.x', '5.0.100' is installed - *skips installation*). I also don't want to speak as something given due to the fact I'm just a random contributor, so it's better to take GH Actions team words for mine but I do remember .NET being nightmare to maintain due to various issues (like installation time which I tried to help mitigate but for Ubuntu, ref: https://github.com/actions/virtual-environments/discussions/2367).

> Totally understand it takes time to ensure everything works it was more an observation of the fact that `-latest` implies a little risk because you're asking for the latest

> Another possible option would be to

use `windows-current`, it doesn't imply any (in)stability and/or version of software installed on the runner
although, I'm sure if that was or were to be used as the name it would confuse new/old users "why it's not latest anymore" and some could find `current` as inadequate choice of word for the tag (somehow), nature of language and conveying message in as few words as possible

**Comment by stap123** [comment]
> use `windows-current`, it doesn't imply any (in)stability and/or version of software installed on the runner although, I'm sure if that was or were to be used as the name it would confuse new/old users "why it's not latest anymore" and some could find `current` as inadequate choice of word for the tag (somehow), nature of language and conveying message in as few words as possible

Naming is hard for sure 🤣

**Comment by Tyler887** [comment]
> @tt2468 innosetup can be installed relatively fast in runtime. Please feel free to ask for help if needed

Would have run `choco install innosetup`.

**Comment by DominusExult** [comment]
> > @tt2468 innosetup can be installed relatively fast in runtime. Please feel free to ask for help if needed
> 
> Would have run `choco install innosetup`.

Is this what @tt2468 meant?

**Comment by jedieaston** [comment]
There are Windows SDKs and libraries (Spectre mitigation is the notable one for me) missing from the VS 2022 installation that were present in the 2019 image, are we to assume these won't be added to this image? I could install them at runtime but that takes a while :(

**Comment by jsoref** [comment]
Personally, I'd suggest `-default`, although I'd be totally happy w/ `-current`.

`-latest` is clearly a misnomer.

**Comment by micdenny** [comment]
Can we disable the warning without the need to update the pipeline yml to stick on `windows-2022`?

**Comment by miketimofeev** [comment]
@micdenny do you use AzureDevOps? If yes — I'm afraid it's not possible at the moment.

**Comment by micdenny** [comment]
@miketimofeev yes we use azuredevops, btw thank you for the reply.

**Comment by drwill-ms** [comment]
Our existing pipeline, updated specifically to windows-2022 and VS 2022 now has an error that sn.exe can't be found trying to disable strong name validation. Suggestions?

**Comment by ssbarnea** [comment]
Can we please get the upcoming change warnings disabled for distributions with `-latest` suffix? Anyone that uses `-latest` already made a decision about what they want to use.

I find perfectly normal to warn people that do not use `latest` about changes that may impact them.

Current behavior add noise to pull requests and distracts from other work.

**Comment by MatthewSteeples** [comment]
@ssbarnea Arguably the message _only_ wants to be shown to people that are using `-latest` as that's the only people that it affects. In our case it was a useful prompt as we then tested our builds against `-2022` on a branch, discovered that they didn't work, and then pinned our builds to `-2019`. Without this warning we'd have been caught off guard when the change happened, and it would have need to have been a rushed and unscheduled change rather than part of a planned sprint.

**Comment by ssbarnea** [comment]
@MatthewSteeples I am inclined to believe that most people will not do anything about it until the change is happening. Testing before change is happening requires human time, and based on what I see on ~100 project which I am watching/contributing is that it is quite uncommon for someone to go an "let's see what happens if I try to update to newer distro".

The reality is that yesterday I seen my builds failing just because Microsoft decided to use pipx to pre-install ansible-core using pipx instead of using the outdated ansible from Ubuntu. That was a change on the image and I do not blame them for doing that change, jut is an example that this warning is questionable and we should have a way to disable it. 

It should not be so hard to add an entry to yaml to declare that you don't want to be notified about new releases. In the end, I am not aware of subscribing to a changelog to build images either, and I would find that far more useful than "asking user to test our new stuff". At least with email, i would be able to decide if latest changes worth testing or not and do not pile up on each CI run.

Should I also mention that the `... soon` is extremely vague, is can be today, next week, next month on or even later. Nobody knows and we are all forced to see that "advertisement" annotation that we cannot control.

Just this morning I had to dig on the issue tracker to find an undocumented feature to disable warning when archive artifact does not have anything to archive. `if-no-files-found: ignore` was the solution, nice and simple.

I am only asking not to punish those that want to use/support/test with the latest version of a specific platform. They already made that statement by picking a label like that. Those that value stability more, can use specific versions.

**Comment by jsoref** [comment]
The workaround today is to pin forward or pin backward. This is of course annoying because in either case one gets commit churn and in either case one has to unpin at some later date.

It'd certainly be nice if GitHub supported a repository/org secret that says `DO_NOT_WARN_IF_WINDOWS_LATEST_IS_2022` with a specific cookie value (to be announced in a task like this one).

Relying on a cookie value would be necessary to prevent people from proactively setting `DO_NOT_WARN_IF_WINDOWS_LATEST_IS_2023`, `DO_NOT_WARN_IF_WINDOWS_LATEST_IS_2024`,  ... `DO_NOT_WARN_IF_WINDOWS_LATEST_IS_2038`.

**Comment by Joren-Thijs-KasparSolutions** [comment]
Are there any issues tracking the update progress on the ubuntu-latest image?

I agree that the label `-latest` should alweays be the latest version available and not something we will update 4 to 5 months down the line...

For example my team has been having a lot of trouble moving our microservices from Net 5 to Net 6 because it's currently unsupported in the "latest" images. Forcing us to run our CI steps in containers which brings it's own problems.

Furthermore i just don't find it acceptable making us wait 4 to 5 months before we can upgrade to a new version when github has it available on day one. especially with Net 6 being an LTS.
The same went for XCode 13 support on MacOS, we just went months without building or testing our IOS version of our flutter app because we simply couldn't, while again github had it on day one.

Please introduce a actual "latest" image so people who want to move forward can move forward.

**Comment by ssbarnea** [comment]
I feel @Joren-Thijs-KasparSolutions pain but I think the reason why we are on this unfortunate situation is because of a past naming mistake that was done.

My guess is that the team producing the system images imagined that they can easily release a fully working image soon after an operating system was released. The reality is that these images are, for good reasons, a kitchensink of tools and it would be close to impossible to get all of them compatible with newer operating system in time.

You cannot really expect to pass all the test suites on a new operating system, some changes will break some of the (build) tools on these platform, and I think that is something we should only get used to.

I do remember seeing delays of even more than 6 months in getting updates in the past but I seen some speed improvements too, it gets better.

Still, how about just updating the documentation in order to state that `-latest` counts as unstable and that many features are not guaranteed to work. If you want you can even add a new set of labels with `-stable` suffix which would point to versions of the operating system that passed the test of time.

Example of less confusing setup:
* `windows-2022` no confusions about what it represents, best of **risk-sensitive** ones
* `windows-stable` an image change that points to a well tests/supported version of the platform. Newer may exist but they would likely have known bugs.
* `windows-latest` point to newest available version of windows as a runner, even if that one may not contain or work with with all build tools. That is aimed to **risk-takers**

This is also a common naming pattern for container images so nobody should be confused by it.

**Comment by rsouthworth** [comment]
The following step, [described here](https://docs.microsoft.com/en-us/sql/tools/sqlpackage/sqlpackage-pipelines?view=sql-server-ver15), might be causing an error:

- script: sqlpackage.exe /version
  workingDirectory: C:\Program Files\Microsoft SQL Server\150\DAC\bin\
  displayName: 'get sqlpackage version'
  
  ##[error]Container path not found: 'C:\Program Files\Microsoft SQL Server\150\DAC\bin\'

**Comment by miketimofeev** [comment]
@rsouthworth the path has changed to the `C:\Program Files\Microsoft SQL Server\160\DAC\bin`. Sorry for the inconvenience.
I'll take a look at updating the docs.

**Comment by drakkan** [comment]
Hi,

Inno Setup removal will break my Windows workflows, see for example [here](https://github.com/drakkan/sftpgo/blob/main/.github/workflows/release.yml#L161). 

I guess I can download and install it as part of my workflows. The latest version of Inno Setup appears to have nearly 800,000 downloads, so its popularity doesn't seem that low. I'm just wondering if you could reconsider this decision, thank you

**Comment by dechamps** [comment]
I also use Inno Setup in my workflows. It's not clear to me how the "low popularity" of Inno Setup was ascertained or what the "maintenance concerns" are, but consider this another vote to keep it alive.

**Comment by miketimofeev** [comment]
@drakkan @dechamps would you mind opening an issue to add InnoSetup to the windows-2022 image?

**Comment by drakkan** [comment]
> @drakkan @dechamps would you mind opening an issue to add InnoSetup to the windows-2022 image?

done, #5006 thanks

**Comment by bormm** [comment]
Choosing "-latest" for something means all your builds are not reproducible/deterministic at all. (https://reproducible-builds.org/) This will maybe not even possible with "windows-<version>", because things might are updated even on these machines.

But I agree with @ssbarnea: Everyone selecting "latest" chose the hottest available way already. It does not make a big difference if this is more or less tested by Microsoft or wait some months to allow people to prepare. Because people that wanted to prepare the switch already have a 2022 branch running and switch by merge that to their productive branch if working. All others maybe also chose "latest" because they don't care or never thought about these things. At the end its the responsibility of the development team to choose the right ingredient for the software.

**Comment by jcwchen** [comment]
Hi, thank you for the awesome work.

I would like to confirm: Is windows-latest workflows now using Windows Server 2022 on GitHub Actions? (I saw it was Windows Server 2019 yesterday, but it changes to use Windows Server 2022 just now)

**Comment by drwill-ms** [comment]
> Our existing pipeline, updated specifically to windows-2022 and VS 2022 now has an error that sn.exe can't be found trying to disable strong name validation. Suggestions?
> 
> ```
> ##[debug]INPUT_SCRIPT: 'call "C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\Enterprise\\Common7\\Tools\\VsDevCmd.bat"
> ##[debug]sn -Vr *,31bf3856ad364e35'
> ##[debug]INPUT_WORKINGDIRECTORY: 'D:\a\1\s'
> ##[debug]Asserting container path exists: 'D:\a\1\s'
> Generating script.
> ##[debug]AGENT_VERSION: '2.198.2'
> ##[debug]AGENT_TEMPDIRECTORY: 'D:\a\_temp'
> ##[debug]Asserting container path exists: 'D:\a\_temp'
> ##[debug]Asserting leaf path exists: 'C:\Windows\system32\cmd.exe'
> ========================== Starting Command Output ===========================
> ##[debug]Entering Invoke-VstsTool.
> ##[debug] Arguments: '/D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\304ba344-c007-433b-8fad-d7ce519fe037.cmd""'
> ##[debug] FileName: 'C:\Windows\system32\cmd.exe'
> ##[debug] WorkingDirectory: 'D:\a\1\s'
> "C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\304ba344-c007-433b-8fad-d7ce519fe037.cmd""
> The system cannot find the path specified.
> 'sn' is not recognized as an internal or external command,
> operable program or batch file.
> ##[debug]Exit code: 1
> ##[debug]Leaving Invoke-VstsTool.
> ##[error]Cmd.exe exited with code '1'.
> ```

Any idea how I solve the problem of missing `sn.exe`?

**Comment by LVMVRQUXL** [comment]
> Hi, thank you for the awesome work.
> 
> I would like to confirm: Is windows-latest workflows now using Windows Server 2022 on GitHub Actions? (I saw it was Windows Server 2019 yesterday, but it changes to use Windows Server 2022 just now)

As @miketimofeev said in the issue description, Window Server 2022 is available progressively on GitHub Actions since January, 17. They plan to complete this migration by March, 6.
My workflows are still using Windows Server 2019, so you are from the lucky ones. 😉

**Comment by BatmanGamingYT** [comment]
![Screenshot 2022-02-14 10 54 07 AM](https://user-images.githubusercontent.com/92596501/153898431-29847869-f15d-4637-a078-5c05cd78d340.png)
how do i fix this?

**Comment by alkino** [comment]
Should be close by #5093

**Comment by Y-LyN-10** [comment]
It looks-like Windows 2022 is **already** the `windows-latest` version. Our builds started to fail with the latest available on GitHub version of Node.js (v16.13.2) at `npm install` command.
...
gyp ERR! find VS msvs_version not set from command line or npm config
gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
gyp ERR! find VS unknown version "undefined" found at "C:\Program Files\Microsoft Visual Studio\2022\Enterprise"
gyp ERR! find VS could not find a version of Visual Studio 2017 or newer to use
gyp ERR! find VS looking for Visual Studio 2015
gyp ERR! find VS - not found
gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
gyp ERR! find VS 
gyp ERR! find VS **************************************************************
gyp ERR! find VS You need to install the latest version of Visual Studio
gyp ERR! find VS including the "Desktop development with C++" workload.
gyp ERR! find VS For more information consult the documentation at:
gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
gyp ERR! find VS **************************************************************
gyp ERR! find VS 
gyp ERR! configure error 
gyp ERR! stack Error: Could not find any Visual Studio installation to use
The issue could be in the Node.js version
```

Here is the **successful** output with Windows-2019:
```
...
gyp info find VS using VS2019 (16.11.32106.194) found at:
gyp info find VS "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise"
gyp info find VS run with --verbose for detailed information
gyp info spawn C:\hostedtoolcache\windows\Python\3.7.9\x64\python3.exe
...
```

Looking at the [Readme](https://github.com/actions/virtual-environments/blob/win22/20220207.1/images/win/Windows2022-Readme.md), Microsoft Visual Studio 2022 should be there in the image 🤔

**Comment by jedieaston** [comment]
The version of npm in the image (8.1.2) includes a version of node-gyp that doesn't support vs2022 yet, so it looks for a maximum of 2019.

If you upgrade npm to [v8.1.4](https://github.com/npm/cli/releases/tag/v8.1.4) or later it will work, so I presume you could just do `npm upgrade -g npm@latest` (presuming npm is separate from Node on the image). If you want to upgrade node-gyp separately of npm, [it's a bit more involved](https://github.com/nodejs/node-gyp/blob/master/docs/Updating-npm-bundled-node-gyp.md).

**Comment by bormm** [comment]
> It looks-like Windows 2022 is **already** the `windows-latest` version. 

Why you highlight the **already**? The post mentions clearly "This change will be rolled out [...] beginning on January, 17. We plan to complete the migration by March, 6." We have 17 February.

Regarding the "Microsoft Visual Studio 2022 should be there in the image" I see a "C:\Program Files\Microsoft Visual Studio\2022\Enterprise" in your message also. So yes it should and is there.

For me looks like related to https://github.com/nodejs/node-gyp/issues/2552 but not the the image itself.

**Comment by ssbarnea** [comment]
After discovering about a week ago that my WSL pipeline gets stuck I was able to narrow down the culprit: `windows-2022` runner is no longer able to use WSL1, while `windows-2019` was able to run in few minutes. [execution-example](https://github.com/ansible-community/ansible-lint/runs/5244664910?check_suite_focus=true)

The way it behaves is very weird, it gets so slow that it effectively times-out. That might have something to do with the fact that my codebase is on the `/mnt/c` and that this has a performance impact.

To get an idea about what kind of slowdown we are talking about, the 2019 jobs runs in less than 3 minutes, while 2022 times out after 30 minutes. Even increasing the timeouts does not help because it will likely take 2-3h+ to finish at its current speed.

The stuck happens while running tests with python (tox and a little bit of ansible syntax check as part of the tests).

Do we need a separated bug for this or we already have one?

**Comment by miketimofeev** [comment]
@ssbarnea we haven't heard about such an issue. Feel free to open a new one and please make sure to provide some repro.

**Comment by seidnerj** [comment]
I've had some tests randomly start to timeout with "windows-latest", though no relevant code changes were made. Every time I would re-run the workflow it would be a different test that timed out. I suspected it is related to the switch to windows 2022, once I've set the version to "windows-2019" things seem to be back to normal.

**Comment by ssbarnea** [comment]
> @ssbarnea we haven't heard about such an issue. Feel free to open a new one and please make sure to provide some repro.

Something is clearly borked around windows-2022. While trying to make a simplified repo and workflow to reproduce the problems I ended up with https://github.com/actions/checkout/issues/702 -- please read it and check the outcomes.

**Comment by usjpaq** [comment]
The docs (https://github.com/actions/virtual-environments) still indicate that windows-latest is 2019, which is what led me here. Can that be brought up to date?

**Comment by jsoref** [comment]
@usjpaq, see https://github.com/actions/virtual-environments/issues/4856#issuecomment-1034116799 and https://github.com/actions/virtual-environments/pull/5093#issuecomment-1042661742

They historically have been unwilling to change the documentation to say something like:

> _platform_-latest is currently transitioning from _platform_-_previous_ to _platform_-_next_ and thus your builds may run on either; if you need deterministic behavior, you can specify either in the interim.

The docs team is fairly reluctant to make changes that only temporarily describe a state. However, on rare occasions it has been possible to convince them.

**Comment by ArchieCoder** [comment]
⚠️ Did the VM move to Windows-2022 "today"?

My pure UWP/XAML project does not compile anymore. One project generates 41 errors. It works perfectly locally.

Is there a way to force the workflow on the previous working VM? Our process is stopped :/

[15_Build the sideload solution.txt](https://github.com/actions/virtual-environments/files/8118834/15_Build.the.sideload.solution.txt)

**Comment by bormm** [comment]
> ⚠️ Did the VM move to Windows-2022 "today"?
No the migration has ~~started~~ a month ago. Just as written in the initial post at the top of the discussion.
So the change is of course expected and no surprise.

> Is there a way to force the workflow on the previous working VM? Our process is stopped :/
Of course and yes, by using windows-2019 instead windows-latest. Also as written in the initial post just at the top of the discussion.

**Comment by ArchieCoder** [comment]
@bormm My bad for not reading correctly the top post. After a discussion with Tim and Lance, apparently I need to install the UWP sdk of the specific versions in the workflow.

**Comment by timheuer** [comment]
> Visual Studio 2019 is now gone. We were using the direct path to Msbuild which is now missing.

Check out microsoft/setup-msbuild action to use

**Comment by StingyJack** [comment]
> Sorry for the inconvenience, this is a known issue and will be fixed next week.

@miketimofeev  - Still seeing this when using windows-latest. 
![image](https://user-images.githubusercontent.com/15381181/155219275-4a2820b0-fcf3-446f-8341-9b8779d1a0c4.png)

If I'm using windows-latest then I recognize that I want the latest and that that will change at times. I dont need to have a warning showing up in all the pipelines for this. Can you please just turn it off and find some other way to share this info that does not create a warning?

**Comment by NoahAndrews** [comment]
> > Sorry for the inconvenience, this is a known issue and will be fixed next week.
> 
> @miketimofeev - Still seeing this when using windows-latest. ![image](https://user-images.githubusercontent.com/15381181/155219275-4a2820b0-fcf3-446f-8341-9b8779d1a0c4.png)
> 
> If I'm using windows-latest then I recognize that I want the latest and that that will change at times. I dont need to have a warning showing up in all the pipelines for this. Can you please just turn it off and find some other way to share this info that does not create a warning?

The issue that was fixed was that the warning was showing up when `windows-2019` was specified. Showing the warning for `windows-latest` was critical in figuring out why our build broke unexpectedly. There's no good reason to remove that.

**Comment by miketimofeev** [comment]
**GitHub migration completed!**
Azure DevOps one is delayed and will be completed by the end of March or even the beginning of April.

**Comment by GilesBathgate** [comment]
@stijnherreman I don't understand the complaints about what `windows-latest` means. Surely it means the latest windows environment that github actions support, which github are at their own discretion regarding the timing of. It has nothing to do with the latest version of windows that exists, and I've never come across any CI platform that roll out an Image on day-zero of GA. 🤷

**Comment by sanking23** [comment]
@sanking23

**Comment by stijnherreman** [comment]
@GilesBathgate

> I've never come across any CI platform that roll out an Image on day-zero of GA.

Neither have I, but we are not talking about day-zero.

> Surely it means the latest windows environment that github actions support, [...]

Azure DevOps in my case, but yes, exactly! But even though `windows-2022` has been available and supported for at least 3+ months, `windows-latest` points at an older version. Plenty of other people at https://github.com/dotnet/core/issues/6907 are also surprised by it.

**Comment by GilesBathgate** [comment]
BTW @miketimofeev It looks like this page needs updating (https://github.com/actions/virtual-environments)

> Neither have I, but we are not talking about day-zero.

@stijnherreman Granted you might not be talking about day-zero, but at least one person was. (although visual-studio not windows)

> Why this was not done on the day of formally releasing visual studio 2022?

The thing is I think some people (including myself) just put `*-latest` in a fire and forget sort of way. As such they don't want to actively change it windows-2022 and so it makes sense for that to be an experimental non stable version. When the environment is considered stable it makes sense to move latest over. As some have said `windows-stable` might be a better name, but a name is just a name, what it means should be documented: https://github.com/actions/virtual-environments

**Comment by StingyJack** [comment]
> Showing the warning for `windows-latest` was critical in figuring out why our build broke unexpectedly. Removing the highly-visible warning would be a mistake IMO.

@NoahAndrews 

Relying on the hosted images Microsoft provides as being stable is the mistake, but I know where you are coming from. I've opened many bug reports and general complaints about my build just starting to break one day because the maintainers did something like dropped python version 3.7 from the windows-2019 image without notice (and without updating the image inventory list for weeks after it had been done).  Or increasing the version of nodejs by _two major versions_ without them even realizing it had happened. The hosts that MS provides are never going to have a fixed version of anything, especially when the image label is "latest". (Links to a few of these complaints that havent been wiped from GH: [here](https://github.com/actions/virtual-environments/issues/1260) and [here](https://github.com/actions/virtual-environments/issues/1526) and [here](https://github.com/actions/virtual-environments/issues/281#issuecomment-602229033) )

What I've learned form these experiences and the response to them is primarily that if a particular pipeline must always have a stable build environment, I need to supply and control the host agent.  And if I can live with a slightly wobbly host agent using the "latest" versions of stuff, I dont need to have a silly warning on every build execution for several months telling me that the image is changing. Its repetitive and obvious.

**Comment by NoahAndrews** [comment]
Oh yeah, I totally agree. I think this transition was handled completely fine. I was just saying that getting rid of the warning would be a big step backwards.

**Comment by ssbarnea** [comment]
If anyone observed a regression from windows-2019 to windows-2022 where the runner stops responding but the job/step still display are running until it eventually reach the job-timeout timeout, please check https://github.com/actions/virtual-environments/issues/5151

No logs are recorded at all and I can ensure that when this happens network connectivity with the runner is totally lost as I has tmate debugging session and they get dropped too right away.

Maybe unrelated to this but I also noted that running a `ping google.com` is a guarantee way kill a runner. It seems that in less than 10s your runner gets killed (not stuck like bug mentioned above). Localhost works but is not a reliable way to assert network connectivity of a machine.

**Comment by timheuer** [comment]
@HuguesJ check the included software on the images. My guess is the SDK and/or toolset version you need is no longer on the `windows-2022` images

**Comment by trayseaof** [comment]
I hope

**Comment by StarOnTheSky** [comment]
I found a strange problem.
I have a CMake workflow using Visual Studio. On Windows Server 2022, if I use `Visual Studio 16 2019`, the x64 target ( `X64`) cannot be built (CMake cannot find the toolchain). But if I use `Visual Studio 17 2022`, the x86 target (`Win32`) cannot be built (toolchain not found, too). 
My workflow definition:
```
        - name: "Windows x64"
          os: windows-latest
          cmake-args: -G "Visual Studio 17 2022" -A x64
          package-file: mmoteeworlds-win64.zip
          package-full-file: mmoteeworlds-win64-full.zip
          env:
            CFLAGS: /WX
            CXXFLAGS: /WX
            LDFLAGS: /WX

        - name: "Windows x86"
          os: windows-2019
          cmake-args: -G "Visual Studio 17 2022" -A Win32
          package-file: mmoteeworlds-win32.zip
          package-full-file: mmoteeworlds-win32-full.zip
          env:
            CFLAGS: /WX
            CXXFLAGS: /WX
            LDFLAGS: /WX
```
CMake error output:
```
CMake Error at CMakeLists.txt:20 (project):
  Generator

    Visual Studio 17 2022

  could not find any instance of Visual Studio.



-- Configuring incomplete, errors occurred!
See also "D:/a/Teeworlds-MRPG/Teeworlds-MRPG/release/CMakeFiles/CMakeOutput.log".
CMake Error:
  Generator

    Visual Studio 17 2022

  could not find any instance of Visual Studio.



CMake Error:
  Generator

    Visual Studio 17 2022

  could not find any instance of Visual Studio.
```

**Comment by remyjette** [comment]
@StarOnTheSky Neither image contains both versions of Visual Studio.

`windows-2019` contains Visual Studio 2019. So you'd use the CMake generator `"Visual Studio 16 2019"` on that image.

`windows-2022` contains Visual Studio 2022. So you'd use the CMake generator `"Visual Studio 17 2022"` on that image.

**Comment by StarOnTheSky** [comment]
> @StarOnTheSky Neither image contains both versions of Visual Studio.
> 
> `windows-2019` contains Visual Studio 2019. So you'd use the CMake generator `"Visual Studio 16 2019"` on that image.
> 
> `windows-2022` contains Visual Studio 2022. So you'd use the CMake generator `"Visual Studio 17 2022"` on that image.

Oh, sorry. It's a silly typo of mine. Thanks.

**Comment by mabmalia** [comment]
It's not possible to build Analysis Services Projects using windows-2022 (only with windows-2019):
`error MSB4226: The imported project "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Business Intelligence Semantic Model\1.0\Microsoft.AnalysisServices.VSHostBuilder.targets" was not found. Also, tried to find "Business Intelligence Semantic Model\1.0\Microsoft.AnalysisServices.VSHostBuilder.targets" in the fallback search path(s) for $(MSBuildExtensionsPath) - "C:\Program Files (x86)\MSBuild" . These search paths are defined in "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\msbuild.exe.Config". Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths`
Does this relates with the fact that the extensions for Analysis Services, Integration Services, and Reporting Services projects are not available for Visual Studio 2022 at this time?

**Comment by allendewilzin** [comment]
> <img alt="image" width="493" src="https://user-images.githubusercontent.com/350314/150398675-f58a3fc9-643e-4dfc-bb24-5fbeb6424321.png">
> 
> What about LTS .NET 6.0?

Where did you get this? @neman

**Comment by mnabe** [comment]
@allendewilzin It is in the issue description.

**Comment by miketimofeev** [comment]
> It's not possible to build Analysis Services Projects using windows-2022 (only with windows-2019): `error MSB4226: The imported project "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Business Intelligence Semantic Model\1.0\Microsoft.AnalysisServices.VSHostBuilder.targets" was not found. Also, tried to find "Business Intelligence Semantic Model\1.0\Microsoft.AnalysisServices.VSHostBuilder.targets" in the fallback search path(s) for $(MSBuildExtensionsPath) - "C:\Program Files (x86)\MSBuild" . These search paths are defined in "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\msbuild.exe.Config". Confirm that the path in the <Import> declaration is correct, and that the file exists on disk in one of the search paths` Does this relates with the fact that the extensions for Analysis Services, Integration Services, and Reporting Services projects are not available for Visual Studio 2022 at this time?

yes, they are not available for VS2022 at the moment https://docs.microsoft.com/en-us/sql/ssdt/download-sql-server-data-tools-ssdt?view=sql-server-ver15#changes-in-ssdt-for-visual-studio-2022

**Comment by sdecoodt** [comment]
> > That's a fair point, on the flip side would it not make sense to have no .NET versions available on the image by default so that it's not possible for people to do it wrong? This would force people down the happy path when creating pipelines.
> 
> Some versions of software that can be installed through `actions/setup-*` is kept in the image because of popularity (of version and not only) and (not sure if all since I didn't check all actions but) `setup-*` tasks will match against requested version and if it matches, it will skip installation (e.g. task requires version '5.0.x', '5.0.100' is installed - _skips installation_). I also don't want to speak as something given due to the fact I'm just a random contributor, so it's better to take GH Actions team words for mine but I do remember .NET being nightmare to maintain due to various issues (like installation time which I tried to help mitigate but for Ubuntu, ref: #2367).
> 
> > Totally understand it takes time to ensure everything works it was more an observation of the fact that `-latest` implies a little risk because you're asking for the latest
> 
> > Another possible option would be to
> 
> use `windows-current`, it doesn't imply any (in)stability and/or version of software installed on the runner although, I'm sure if that was or were to be used as the name it would confuse new/old users "why it's not latest anymore" and some could find `current` as inadequate choice of word for the tag (somehow), nature of language and conveying message in as few words as possible

Two words might just be inadequate. 
It's the current default version of windows determined by the team :) windows-default / windows-current-default ...

**Comment by miketimofeev** [comment]
The migration is completed for everyone. The docs will be updated soon.

**Comment by timheuer** [comment]
@zachpainter77 change your agents to `windows-2019` and your scenario should be back to good.

**Comment by Mike-Bell** [comment]
This might not be the appropriate place to post this, but FWIW this warning message is now outdated and confusing (an updated warning of "windows-latest pipelines recently switched to use windows-2022" might have helped save me a little time figuring out why my build broke over night).

![image](https://user-images.githubusercontent.com/8780033/160869181-5d09402a-bdc5-4b82-b79f-2624cc0a2a26.png)

**Comment by zachpainter77** [comment]
> @zachpainter77 change your agents to `windows-2019` and your scenario should be back to good.

@timheuer  Thank you, we did this and all is working.

**Comment by miketimofeev** [comment]
> This might not be the appropriate place to post this, but FWIW this warning message is now outdated and confusing (an updated warning of "windows-latest pipelines recently switched to use windows-2022" might have helped save me a little time figuring out why my build broke over night).
> 
> ![image](https://user-images.githubusercontent.com/8780033/160869181-5d09402a-bdc5-4b82-b79f-2624cc0a2a26.png)

Sorry to hear that. We are going to change the message soon.

**Comment by JanRajnoha** [comment]
I have problem with building Windows App SDK on Windows 2022. 2019 works fine.

Error:
C:\hostedtoolcache\windows\dotnet\sdk\6.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(427,5): Error NETSDK1112: The runtime pack for Microsoft.Windows.SDK.NET.Ref was not downloaded. Try running a NuGet restore with the RuntimeIdentifier 'any'.

Is there any difference or did I missing something?

**Comment by timheuer** [comment]
@JanRajnoha likely because the target Windows SDK is not on the image, you can see the versions:
windows-2022: [https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md?rgh-link-date=2022-03-23T16%3A49%3A14Z)
windows-2019: [https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md?rgh-link-date=2022-03-23T16%3A49%3A14Z)

**Comment by JanRajnoha** [comment]
@timheuer I'm using only 22000 SDK, I checked both VMs and are in all specific parameters same :/

**Comment by timheuer** [comment]
@JanRajnoha I'd recommend switching back to `windows-2019` of course while you debug to get you passing.  Might be good to open a new separate issue on this one.  My guess is that this is just due to a missing component that is no longer in the image.  If the Windows SDK you need is present it may be relying on an underlying ref pack (given the error) that points to `windows-2022` not having that version of the .NET Framework -- you can try to enable this by getting it on the image if you can identify what is missing...or switching to 2019.

**Comment by miketimofeev** [comment]
> @miketimofeev how do we remove this warning? it causes our builds to fail as it generates a warning.

the warning is in place for the last two months. Does it cause your builds to fail all that time?
We are going to remove it next week.

**Comment by jozefizso** [comment]
Yes it does as we have 0 warning policy.

**Comment by jsoref** [comment]
@retchegaraydev: In theory `windows-2016`, was scheduled to be removed on March 15, 2022: https://github.blog/changelog/2021-10-19-github-actions-the-windows-2016-runner-image-will-be-removed-from-github-hosted-runners-on-march-15-2022/
That didn't quite happen: https://github.com/actions/virtual-environments/issues/5238 -- at present (Apr 3), windows-2016 appears to be unavailable -- although it's possible it's another brownout as opposed to fully removed (certainly lots of the documentation and validation is still at the "it's not quite dead yet" stage).

While it seemed likely that they'd remove windows-2019 runners around the time that they make available the replacement for windows-2022, in theory they might do it earlier. Speculation seems to hint that a Windows Server 2025 could happen. But the [EOL for windows-2019](https://docs.microsoft.com/en-us/lifecycle/products/windows-server-2019) is Jan 2024.

I think it's likely that they'd drop windows-2019 around Mar 2024/Apr 2024 just as they dropped windows-2016 a couple of months after it reached [EOL](https://docs.microsoft.com/en-us/lifecycle/products/windows-server-2016).

**Comment by sachinbdoshi** [comment]
I have a question with regards to our containers that we build. the containers required AzureML SDK and the latest python verion that it supports python 3.8 https://pypi.org/project/azureml-sdk/?msclkid=23fb8cfab55a11ecb682625f469f2675 the table published on this page mentions that win 2022 has python 3.9. so can I safely assume that It will not build our containers in the Azure Pipeline and we will have to stick to using windows-2019 as our agent pool ?

**Comment by miketimofeev** [comment]
@sachinbdoshi python 3.9 is set as the default version but there are many other versions pre-installed, just use https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/use-python-version?view=azure-devops task
<img width="354" alt="image" src="https://user-images.githubusercontent.com/48208649/161933967-91d2bd38-04c9-4844-b11f-71a887597877.png">

**Comment by sachinbdoshi** [comment]
Thank you, I will gove this option a try.

Regards,
Sachin Doshi

**Comment by miketimofeev** [comment]
@jozefizso the message has been disabled.

**Comment by donatellijl** [comment]
It seems that windows-latest is still pointing to windows-2019 instead of windows-2022. When I run my pipeline in Azure DevOps with the vmImage set to windows-latest, I see that the pool is set to windows-2019.

![Screen Shot 2022-10-04 at 7 53 44 AM](https://user-images.githubusercontent.com/45564887/193813145-c43b3d4b-ea40-4cee-88b6-ffc10095c194.png)

When I change to windows-2022, the pool is set to windows-2022.

![Screen Shot 2022-10-04 at 7 55 13 AM](https://user-images.githubusercontent.com/45564887/193813306-dcd588fa-4e9e-4dec-af96-b5be1accfe37.png)

According to the [docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml), windows-latest should be windows-2022. 

I'm also confused as to why the docs say that for classic pipelines, the default is windows-2019. Shouldn't it be windows-latest?
![Screen Shot 2022-10-04 at 7 58 51 AM](https://user-images.githubusercontent.com/45564887/193813878-e3d66e08-2e52-402a-a229-d7d519a5f6d1.png)

**Comment by al-cheb** [comment]
@donatellijl, Could you check?
```
pool:
  name: Azure Pipelines
  vmImage: windows-latest
```

**Comment by donatellijl** [comment]
> @donatellijl, Could you check?
> 
> ```
> pool:
>   name: Azure Pipelines
>   vmImage: windows-latest
> ```

I added "name: Azure Pipelines" and now instead of my first screenshot, I see windows-latest as the Image. Why is having the name a requirement?

![Screen Shot 2022-10-04 at 8 32 15 AM](https://user-images.githubusercontent.com/45564887/193820611-899bb1a9-38bc-401f-a784-354dc7a47050.png)

**Comment by VikasNischalDA** [comment]
Hello everyone, I hope you all are well.
I was trying to create a database with Microsoft hosted agent in yml pipeline. Can anyone please tell me if these agents offers any database server? This will be a great help 👍

**Comment by princekumarLnG** [comment]
Hello everyone, I hope you all are well.
I was trying to create a database with Microsoft hosted agent in yml pipeline. Can anyone please tell me if these agents offers any database server? This will be a great help 👍

**Comment by timheuer** [comment]
> Hello everyone, I hope you all are well. I was trying to create a database with Microsoft hosted agent in yml pipeline. Can anyone please tell me if these agents offers any database server? This will be a great help 👍

Complete list of software is available at the image definitions: https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md (for `windows-latest`)
